content/blog/2025-05-13-abstract-testing-floating-point.md

```diff
@@ -0,0 +1,201 @@
++++
+title = "What is a bug? New ways to test verification tools."
+[extra]
+latex = true
+bio = "Max Fan is a first-year CS PhD student at Cornell University."
+[[extra.authors]]
+name = "Max Fan"
+link = "https://max.fan"
++++
+
+My project is driven by a deceptively simple question: *What is a bug in a
+program verification tool?* 
+
+In the following blog post, I will:
+1. Explore a relaxed definition of a bug that ought to enable more effective
+   testing of verification tools. 
+2. Outline how to apply my definition to _abstractly test_ floating-point
+   verification software.
+3. Discuss my prototype implementation and preliminary results.
+
+```
+                                _______
+                               /       \
+                            ---|       |---
+                               |  Bug  |
+                            ---|       |---
+                               \_______/
+                                /  |  \
+                               /   |   \
+                              "    |    "
+```
+
+
+## What is a bug?
+
+Consider a verification tool **VERIFY** to take some program *p(x)* and provide
+a **GUARANTEE** such that for all inputs *x*, **GUARANTEE(**_p(x)_**)** holds.
+At first blush, one might define a bug as follows: 
+
+> <ins>Definition</ins> (_bug_). A bug in **VERIFY** exists iff there exists
+an input _x_ such that **GUARANTEE(**_p(x)_**)** does not hold.
+
+This is a perfectly reasonable definition. If you can violate a guarantee, it
+seems like you have a bug.
+
+But deep down in our hearts, there is _secretly_ a stronger notion of
+correctness that we'd like the verification tool to respect. In particular, it
+would be really bad if _I_ published a paper claiming my tool does **X** and
+prove **X** to be a sound technique when in reality it does **Y**. That sounds
+like a bug!
+
+In other words: we'd like our verification tool to faithfully implement its own
+theory (and respect its corresponding soundness proof). This is an entirely
+reasonable demand for a user to have. I absolutely want my tool to do what it
+claims to be doing.
+
+Assuming the tool's theory is sound, we can define a new notion of bug, which I
+call an _abstract bug_ (for reasons what will become apparent).
+
+> <ins>Definition</ins> (_abstract bug_). An abstract bug in **VERIFY** exists
+> iff it does not faithfully implement its own theory. 
+
+## Abstract Testing of Floating-Point Software
+
+With a fancy new definition, let's hunt for bugs! Observe that many (but not
+all) verification tools operate over an abstract domain using overapproximate
+simplifying assumptions. 
+
+For example, the "standard model of floating-point error" assumes that each
+floating-point operation $(op_{float} \ x \ y)$ can introduce up to $1*\epsilon$
```

Comment:
Do you mean 1+\epsilon? (That's what appears in the equation below.)

Reply:
Yep, fixed -- good catch. Thanks!